jcat.c

```diff
@@ -0,0 +1,390 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Copyright (c) 2018 Pengutronix, Oleksij Rempel <o.rempel@pengutronix.de>
+ */
+
+#include <err.h>
+#include <errno.h>
+#include <error.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <net/if.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "libj1939.h"
+#define J1939_MAX_ETP_PACKET_SIZE (7 * 0x00ffffff)
+#define JCAT_BUF_SIZE (1000 * 1024)
+
+/*
+ * min()/max()/clamp() macros that also do
+ * strict type-checking.. See the
+ * "unnecessary" pointer comparison.
+ */
+#define min(x, y) ({				\
+	typeof(x) _min1 = (x);			\
+	typeof(y) _min2 = (y);			\
+	(void) (&_min1 == &_min2);		\
+	_min1 < _min2 ? _min1 : _min2; })
+
+struct jcat_priv {
+	int sock;
+	int infile;
+	int outfile;
+	unsigned int todo_send;
+	int todo_prio;
+
+	bool valid_peername;
+	bool todo_recv;
+	bool todo_filesize;
+	bool todo_connect;
+
+	struct sockaddr_can sockname;
+	struct sockaddr_can peername;
+};
+
+static const char help_msg[] =
+	"jcat: netcat tool for j1939\n"
+	"Usage: jcat FROM TO\n"
+	" FROM / TO	- or [IFACE][:[SA][,[PGN][,NAME]]]\n"
+	"Options:\n"
+	" -i <infile>	(default stdin)\n"
+	" -s[=LEN]	Initial send of LEN bytes dummy data\n"
+	" -r		Receive data\n"
+	"\n"
+	"Example:\n"
+	"jcat -i some_file_to_send  can0:0x80 :0x90,0x12300\n"
+	"jcat can0:0x90 -r > /tmp/some_file_to_receive\n"
+	"\n"
+	;
+
+static const char optstring[] = "?i:vs::rp:cnw::";
+
+
+static void jcat_init_sockaddr_can(struct sockaddr_can *sac)
+{
+	sac->can_family = AF_CAN;
+	sac->can_addr.j1939.addr = J1939_NO_ADDR;
+	sac->can_addr.j1939.name = J1939_NO_NAME;
+	sac->can_addr.j1939.pgn = J1939_NO_PGN;
+}
+
+static void parse_canaddr(char *spec, struct sockaddr_can *paddr)
+{
+	char *str;
+
+	str = strsep(&spec, ":");
+	if (strlen(str))
+		paddr->can_ifindex = if_nametoindex(str);
+
+	str = strsep(&spec, ",");
+	if (str && strlen(str))
+		paddr->can_addr.j1939.addr = strtoul(str, NULL, 0);
+
+	str = strsep(&spec, ",");
+	if (str && strlen(str))
+		paddr->can_addr.j1939.pgn = strtoul(str, NULL, 0);
+
+	str = strsep(&spec, ",");
+	if (str && strlen(str))
+		paddr->can_addr.j1939.name = strtoul(str, NULL, 0);
+}
+
+static int jcat_sendfile(struct jcat_priv *priv, int out_fd, int in_fd,
+			 off_t *offset, size_t count)
+{
+	int ret = EXIT_SUCCESS;
+	off_t orig = 0;
+	char *buf;
+	size_t to_read, num_read, num_sent, tot_sent, buf_size;
+	int round = 0;
+
+	buf_size = min((size_t)J1939_MAX_ETP_PACKET_SIZE, count);
+	buf = malloc(buf_size);
+	if (!buf) {
+		warn("can't allocate buf");
+		ret = EXIT_FAILURE;
+		goto do_nofree;
+	}
+
+	if (offset != NULL) {
```

Comment:
`if (!offset)`

Reply:
@olerem pointed out that should be `if (offset)`. doh!.
Fixed with https://github.com/linux-can/can-utils/pull/117